File: slam/localization/src/gps_decoder.py
#!/usr/bin/env python

import rospy
from nmea_msgs.msg import Sentence
from localization.msg import Gps
from pyproj import Proj, transform
#from functools import reduce ## for python3 only

class GPS_Decoder:
    proj_WGS84 = Proj(init = 'epsg:4326')
    proj_5186 = Proj(init = 'epsg:5186') # for FMTC
    proj_UTMK = Proj(init = 'epsg:5178') # for K-city
    proj_UTM52N = Proj(init = 'epsg:32652')
        
    def __init__(self):
        self.pub_ = rospy.Publisher("/gps", Gps, queue_size=10)
        self.sub_ = rospy.Subscriber("/nmea_sentence",Sentence, self.callback)

    def callback(self, msg):
        rt = Gps()
        s = msg.sentence
        tokens = s.split(',')
        if tokens[0] == "$GNGGA":
            for token in tokens[1:9]:
                if token == "":
                    return
            if int(tokens[6])<2:
                rospy.logerr("failedByQualityIndicator(No-Fix), GPS Quality Indicator: %s", tokens[6])
                return
            if int(tokens[6])<4:
                rospy.logwarn("warnByQualityIndicator(No-RTK), GPS Quality Indicator: %s", tokens[6])
            if int(tokens[7])<3:
                rospy.logerr("failedByStatellitesN, The number of Satellite: %s", tokens[7])
                return
            if int(tokens[7])<8:
                rospy.logwarn("warnByStatellitesN, The number of Satellite: %s", tokens[7])
            if (float(tokens[8])>7.5 or float(tokens[8])==0.0):
                rospy.logerr("failedByHDOP, HDOP value: %s", tokens[8])
                return
            if float(tokens[8])>3.5:
                rospy.logwarn("warnByHDOP, HDOP value: %s", tokens[8])
            
            checksum = s[1:].split('*')
            cs0 = reduce(lambda x,y:x^y, map(ord,checksum[0]))
            cs1 = int(checksum[1],16)
            if cs0 != cs1:
                rospy.logerr("FailedByChecksum, %d != %d == 0x%s", cs0, cs1, checksum[1])
                return

            # hour = int(tokens[1][0:2])
            # minu = int(tokens[1][2:4])
            # sec = float(tokens[1][4:])
            # time = hour*3600 + minu*60 + sec
            lat = int(tokens[2][0:2]) + float(tokens[2][2:])/60.0
            lon = int(tokens[4][0:3]) + float(tokens[4][3:])/60.0
            
            rt.header = msg.header
            rt.x, rt.y = transform(self.proj_WGS84,self.proj_5186,lon,lat)
            #rt.header.stamp = rospy.Time(time)
            #rt.header.stamp = rospy.get_rostime()
            rt.err_identifier = float(tokens[6])
            rt.err_hdop = float(tokens[8])
            self.pub_.publish(rt)
        
    

if __name__ == '__main__':
    rospy.init_node('gps_decoder')
    gps_decoder=GPS_Decoder()
    rospy.spin()

File: slam/src/sensing/gps_decoder.cpp
#include "ros/ros.h"
#include "nmea_msgs/Sentence.h"
#include "slam/Gps.h"
#include "std_msgs/Duration.h"
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include "UTM.h"
using namespace std;

double QI_err;
double QI_warn;
double Satellites_err;
double Satellites_warn;
double HDOP_err;
double HDOP_warn;

class GPS_Decoder{
    public:
    GPS_Decoder(){
        pub_ = n_.advertise<slam::Gps>("gps", 2);
        sub_ = n_.subscribe("/nmea_sentence", 2, &GPS_Decoder::callback, this);
        pub_delay = n_.advertise<std_msgs::Duration>("delay_gps", 10000);
    }	

    void callback(const nmea_msgs::Sentence::ConstPtr& msg){
        ros::Time t0 = ros::Time::now();
        slam::Gps rt;
        string s = msg->sentence;
        string delimiter = ",";
        vector<string> tokens;
        string token;
        size_t pos = 0;

        while ((pos = s.find(delimiter)) != std::string::npos) {
            token = s.substr(0, pos);
            tokens.push_back(token);
            s.erase(0, pos + delimiter.length());
        }

	//edited by boseol
        if(tokens[0] == "$GNGGA"){
            double time_raw, lat_raw, lon_raw;
            double time, lat, lon;
            try{
                if(stoi(tokens[6])<QI_err){ROS_ERROR("failedByQualityIndicator(No Satellite), GPS Quality Indicator: %s", tokens[6].c_str()); return;}
                if(stoi(tokens[6])<QI_warn){ROS_WARN("warnByQualityIndicator(Non-RTK), GPS Quality Indicator: %s", tokens[6].c_str());}
                if(stoi(tokens[7])<Satellites_err){ROS_ERROR("failedBySatellitesN, The number of Satellite: %s", tokens[7].c_str()); return;}
                if(stoi(tokens[7])<Satellites_warn){ROS_WARN("warnBySatellitesN, The number of Satellite: %s", tokens[7].c_str());}
                if(stod(tokens[8])>HDOP_err || stod(tokens[8])==0.0){ROS_ERROR("failedByHDOP, HDOP value: %s", tokens[8].c_str()); return;}
                if(stod(tokens[8])>HDOP_warn){ROS_WARN("warnByHDOP, HDOP value: %s", tokens[8].c_str());} 

                {   //check checksum 
                    s = msg->sentence;
                    delimiter = "$";
                    pos = s.find(delimiter);
                    s.erase(0, pos + delimiter.length());
                    delimiter = "*";
                    pos = s.find(delimiter);
                    token = s.substr(0, pos);
                    s.erase(0, pos + delimiter.length());
                    int checksum = 0;
                    for (string::iterator iter = token.begin(); iter != token.end(); ++iter){
                        checksum ^= (int)(*iter);
                    }
                    int c;
                    stringstream ss;
                    ss << std::hex << s;
                    ss >> c;
                    if(checksum != c){ROS_ERROR("FailedByChecksum, %d != %d == 0x%s", checksum, c, s.c_str()); return;}
                }

                // time_raw = stod(tokens[1]);
                lat_raw = stod(tokens[2]);
                lon_raw = stod(tokens[4]);
            }catch(...){
                ROS_ERROR("failedByNotANumber");
                return;
            }
            // int hour = (int)time_raw/10000;
            // int min = (int)time_raw/100 - hour*100;
            // double sec = time_raw -(hour*10000+min*100);
            // time = hour*3600 + min*60 + sec;

            lat = (int)lat_raw/100;
            lon = (int)lon_raw/100;

            lat += (lat_raw-lat*100)/60;
            lon += (lon_raw-lon*100)/60;
            
            vector<double> xy(2);
            LatLonToUTMXY(lat, lon, 52, xy.at(0), xy.at(1));
            
            rt.header = msg->header;
            rt.x = xy.at(0);
            rt.y = xy.at(1);
            //rt.header.stamp.sec= int(time);
            //rt.header.stamp.nsec= (time-rt.header.stamp.sec)*1e9;
            //ros::Time tm = ros::Time::now();
            //rt.header.stamp.sec = tm.sec;
            //rt.header.stamp.nsec = tm.nsec;

            pub_.publish(rt);

            std_msgs::Duration dt;
            dt.data = ros::Time::now() - t0;
            pub_delay.publish(dt);
        }
    }

    private:
    ros::NodeHandle n_;
    ros::Publisher pub_;
    ros::Publisher pub_delay;
    ros::Subscriber sub_;
};


int main(int argc, char** argv) {
    ros::init(argc, argv, "gps_decoder");
	ros::param::get("/QI_err", QI_err);
	ros::param::get("/QI_warn", QI_warn);
	ros::param::get("/Satellites_err", Satellites_err);
	ros::param::get("/Satellites_warn", Satellites_warn);
	ros::param::get("/HDOP_err", HDOP_err);
	ros::param::get("/HDOP_warn", HDOP_warn);
    GPS_Decoder GPSObject;
    ros::spin();
}
